File: ZLaunchAdObjc/ZLaunchAdObjc/ZLaunchAdObjc/ZLaunchAdVC.m
//
//  ZLaunchAdVC.m
//  ZLaunchAdObjc
//
//  Created by mengqingzheng on 2017/8/18.
//  Copyright © 2017年 mengqingzheng. All rights reserved.
//

#import "ZLaunchAdVC.h"
#import "UIImageView+ZLaunchAd.h"

#define DEFAULT_TIME 3

@interface ZLaunchAdVC ()

@property (nonatomic, assign) NSInteger         defaultTime;
@property (nonatomic, strong) UIImageView       *launchImageView;       /// 启动页
@property (nonatomic, strong) UIImageView       *launchAdImgView;       /// 启动广告
@property (nonatomic, strong) UIButton          *skipBtn;               /// 跳过按钮
@property (nonatomic, strong) dispatch_source_t originalTimer;          /// 默认定时器
@property (nonatomic, strong) dispatch_source_t dataTimer;              /// 广告定时器
@property (nonatomic, assign) NSInteger         adDuration;             /// 广告停留时间
@property (nonatomic, strong) CAShapeLayer      *animationLayer;
@property (nonatomic, assign) CGFloat           adViewBottomDistance;
@property (nonatomic, assign) TransitionStyle   transitionStyle;
@property (nonatomic,   copy) ZLaunchAdCompletion    adImgViewClick;
@property (nonatomic, strong) ZLaunchAdConfig   *skipBtnConfig;         /// 配置跳过按钮
@property (nonatomic, strong) UIViewController  *rootViewController;

@end

@implementation ZLaunchAdVC

- (instancetype)initWithDuration: (NSInteger)duration transitionStyle: (TransitionStyle)transitionStyle adBottom: (CGFloat)adBottom rootViewController: (UIViewController *)rootViewController {
    self = [super init];
    if (self) {
        self.defaultTime = duration;
        self.transitionStyle = transitionStyle;
        self.adViewBottomDistance = adBottom;
        self.rootViewController = rootViewController;
    }
    return self;
}

- (void)viewDidLoad {
    [super viewDidLoad];
    self.skipBtnConfig = [[ZLaunchAdConfig alloc]init];
    self.defaultTime = self.defaultTime <= 0 ? DEFAULT_TIME : self.defaultTime;
    self.adViewBottomDistance = self.adViewBottomDistance <= 0 ? 100 : self.adViewBottomDistance;
    [self.view addSubview:self.launchImageView];
    [self startTimer];
}

#pragma mark - 配置广告图
- (void)configNetImageWithURL: (NSString *)url duration: (NSInteger)duration adImgViewClick: (ZLaunchAdCompletion)adImgViewClick {
    
    if ([url isEqualToString:@""]) { return; }
    self.adDuration = duration < 1 ? 1 : duration;
    [self.view addSubview:self.launchAdImgView];
    __weak typeof(self)weakSelf = self;
    [self.launchAdImgView configImageWithUrl:url completion:^{
        
        [weakSelf addSkipBtn];
        
        if (!weakSelf.originalTimer) { return; }
        [weakSelf adStartTimer];
        [UIView animateWithDuration:0.8 animations:^{
            weakSelf.launchAdImgView.alpha = 1;
        }];
        
    }];
    self.adImgViewClick = adImgViewClick;
}
/// 本地图片
- (void)configLocalImageWithImage: (UIImage *)image duration: (NSInteger)duration adImgViewClick: (ZLaunchAdCompletion)adImgViewClick {
    if (image == nil) { return; }
    self.adDuration = duration < 1 ? 1 : duration;
    [self.view addSubview:self.launchAdImgView];
    self.launchAdImgView.image = image;
    
    [self addSkipBtn];
    if (!self.originalTimer) { return; }
    
    [self adStartTimer];
    [UIView animateWithDuration:0.8 animations:^{
        self.launchAdImgView.alpha = 1;
    }];
    self.adImgViewClick = adImgViewClick;
}
/// 本地GIF
- (void)configLocalGifWithName: (NSString *)name duration: (NSInteger)duration adImgViewClick: (ZLaunchAdCompletion)adImgViewClick {
    self.adDuration = duration < 1 ? 1 : duration;
    [self.view addSubview:self.launchAdImgView];
    [self.launchAdImgView gifWithName:name completion:^{
        [self addSkipBtn];
        if (!self.originalTimer) { return; }
        
        [self adStartTimer];
        
        [UIView animateWithDuration:0.8 animations:^{
            self.launchAdImgView.alpha = 1;
        }];
    }];
    self.adImgViewClick = adImgViewClick;
}

- (void)addSkipBtn {
    [self.skipBtn removeFromSuperview];
    if (self.animationLayer) {
        [self.animationLayer removeFromSuperlayer];
        self.animationLayer = nil;
    }
    if (self.skipBtnConfig.skipBtnType != SkipBtnTypeNone) {
        
        if (self.configSkipBtn) {
            self.configSkipBtn(self.skipBtnConfig);
        }
        self.skipBtn.backgroundColor = self.skipBtnConfig.backgroundColor;
        self.skipBtn.titleLabel.font = self.skipBtnConfig.titleFont;
        [self.skipBtn setTitleColor:self.skipBtnConfig.titleColor forState:UIControlStateNormal];
        if (self.skipBtnConfig.skipBtnType == SkipBtnTypeCircle) {
            self.skipBtn.frame = CGRectMake(0, 0, self.skipBtnConfig.height, self.skipBtnConfig.height);
            self.skipBtn.layer.cornerRadius = self.skipBtnConfig.height*0.5;
            [self circleBtnAddLayerWithStrokeColor:self.skipBtnConfig.strokeColor lineWidth:self.skipBtnConfig.lineWidth];
        } else {
            self.skipBtn.frame = CGRectMake(0, 0, self.skipBtnConfig.width, self.skipBtnConfig.height);
            self.skipBtn.layer.cornerRadius = self.skipBtnConfig.cornerRadius;
        }
        self.skipBtn.center = CGPointMake(self.skipBtnConfig.centerX, self.skipBtnConfig.centerY);
        [self.view addSubview: self.skipBtn];
        [self.skipBtn setTitle: self.skipBtnConfig.skipBtnType == SkipBtnTypeTimer ? [NSString stringWithFormat:@"\%ld 跳过",self.adDuration] : @"跳过" forState: UIControlStateNormal];
    }
}
/// 添加动画
- (void)circleBtnAddLayerWithStrokeColor: (UIColor *)strokeColor lineWidth: (CGFloat)lineWidth {
    
    UIBezierPath *bezierPath = [UIBezierPath bezierPathWithOvalInRect:self.skipBtn.bounds];
    self.animationLayer = [[CAShapeLayer alloc]init];
    self.animationLayer.path = bezierPath.CGPath;
    self.animationLayer.lineWidth = lineWidth;
    self.animationLayer.strokeColor = strokeColor.CGColor;
    self.animationLayer.fillColor = [UIColor clearColor].CGColor;
    CABasicAnimation *animation = [CABasicAnimation animationWithKeyPath:@"strokeStart"];
    animation.duration = self.adDuration;
    animation.fromValue = @0;
    animation.toValue = @1;
    [self.animationLayer addAnimation:animation forKey:nil];
    [self.skipBtn.layer addSublayer: self.animationLayer];
}
#pragma mark - 点击事件

- (void)imgViewTapAction:(UITapGestureRecognizer *)sender {
    dispatch_source_cancel(self.dataTimer);
    __strong typeof(self)weakSelf = self;
    [self launchAdVCRemoveCompletion:^{
        __strong typeof(weakSelf) strongWeakSelf = weakSelf;
        dispatch_after(dispatch_time(DISPATCH_TIME_NOW, (int64_t)(0.5 * NSEC_PER_SEC)), dispatch_get_main_queue(), ^{
            if (strongWeakSelf.adImgViewClick) {
                strongWeakSelf.adImgViewClick();
            }
        });
    }];
}

- (void)skipBtnAction {
    if (self.dataTimer) {
        dispatch_source_cancel(self.dataTimer);
    }
    [self launchAdVCRemoveCompletion:nil];
}
#pragma mark -  懒加载
- (UIButton *)skipBtn {
    if (!_skipBtn) {
        _skipBtn = [UIButton buttonWithType:UIButtonTypeCustom];
        [_skipBtn addTarget:self action:@selector(skipBtnAction) forControlEvents:UIControlEventTouchUpInside];
    }
    return _skipBtn;
}
- (UIImageView *)launchImageView {
    if (!_launchImageView) {
        _launchImageView = [[UIImageView alloc]initWithFrame:[UIScreen mainScreen].bounds];
        _launchImageView.image = [self getLaunchImage];
    }
    return _launchImageView;
}
- (UIImageView *)launchAdImgView {
    if (!_launchAdImgView) {
        _launchAdImgView = [[UIImageView alloc]initWithFrame:CGRectMake(0, 0, Z_SCREEN_WIDTH, Z_SCREEN_HEIGHT-self.adViewBottomDistance)];
        _launchAdImgView.userInteractionEnabled = true;
        _launchAdImgView.alpha = 0.2;
        UITapGestureRecognizer *tap = [[UITapGestureRecognizer alloc]initWithTarget:self action:@selector(imgViewTapAction:)];
        [_launchAdImgView addGestureRecognizer:tap];
    }
    return _launchAdImgView;
}

#pragma mark - 广告结束
- (void)launchAdVCRemoveCompletion: (ZLaunchAdCompletion)completion {
    if (self.originalTimer) {
        dispatch_source_cancel(self.originalTimer);
    }
    if (self.dataTimer) {
        dispatch_source_cancel(self.dataTimer);
    }
    if (self.rootViewController) {
        [self transitionAnimation];
        [UIApplication sharedApplication].keyWindow.rootViewController = self.rootViewController;
        if (completion) {
            completion();
        }
    }
}
- (void)transitionAnimation {
    CATransition *trans = [[CATransition alloc]init];
    trans.duration = 0.5;
    switch (self.transitionStyle) {
        case TransitionStyleRippleEffect:
            trans.type = @"rippleEffect";
            break;
        case TransitionStyleFlipFromLeft:
            trans.type = @"oglFlip";
            trans.subtype = kCATransitionFromLeft;
            break;
        case TransitionStyleFlipFromRight:
            trans.type = @"oglFlip";
            trans.subtype = kCATransitionFromRight;
            break;
        case TransitionStyleFlipFromTop:
            trans.type = @"oglFlip";
            trans.subtype = kCATransitionFromTop;
            break;
        case TransitionStyleFlipFromBottom:
            trans.type = @"oglFlip";
            trans.subtype = kCATransitionFromTop;
            break;
        default:
            trans.type = @"fade";
            break;
    }
    [[UIApplication sharedApplication].keyWindow.layer addAnimation:trans forKey:nil];
}

#pragma mark - GCD定时器
- (void)startTimer {
    self.originalTimer = dispatch_source_create(DISPATCH_SOURCE_TYPE_TIMER, 0, 0, dispatch_get_global_queue(0, 0));
    dispatch_source_set_timer(self.originalTimer, dispatch_walltime(NULL, 0), 1.0 * NSEC_PER_SEC, 0);
    dispatch_source_set_event_handler(self.originalTimer, ^{
        
        dispatch_async(dispatch_get_main_queue(), ^{
            
            NSLog(@"默认定时器%ld",(long)self.defaultTime);
            
            if (self.defaultTime == 0) {
                [self launchAdVCRemoveCompletion:nil];
            }
            self.defaultTime -= 1;
        });
    });
    dispatch_resume(self.originalTimer);
}

- (void)adStartTimer {
    dispatch_source_cancel(self.originalTimer);
    self.dataTimer = dispatch_source_create(DISPATCH_SOURCE_TYPE_TIMER, 0, 0, dispatch_get_global_queue(0, 0));
    dispatch_source_set_timer(self.dataTimer, dispatch_walltime(NULL, 0), 1.0 * NSEC_PER_SEC, 0);
    dispatch_source_set_event_handler(self.dataTimer, ^{
        
        dispatch_async(dispatch_get_main_queue(), ^{
            
            NSLog(@"广告倒计时%ld",(long)self.adDuration);
            
            [self.skipBtn setTitle: self.skipBtnConfig.skipBtnType == SkipBtnTypeTimer ? [NSString stringWithFormat:@"\%ld 跳过",self.adDuration] : @"跳过" forState: UIControlStateNormal];
            
            if (self.adDuration == 0) {
                [self launchAdVCRemoveCompletion:nil];
            }
            self.adDuration -= 1;
        });
    });
    dispatch_resume(self.dataTimer);
}

#pragma mark - 获取启动页
- (UIImage *)getLaunchImage{
    UIImage *launchImage = [self assetsLaunchImage];
    if(launchImage) return launchImage;
    return [self storyboardLaunchImage];
}

- (UIImage *)assetsLaunchImage{
    CGSize viewSize = [UIScreen mainScreen].bounds.size;
    NSString *viewOrientation = @"Portrait";//横屏"Landscape"
    NSString *launchImageName = nil;
    NSArray* imagesDict = [[[NSBundle mainBundle] infoDictionary] valueForKey:@"UILaunchImages"];
    for (NSDictionary* dict in imagesDict){
        CGSize imageSize = CGSizeFromString(dict[@"UILaunchImageSize"]);
        if (CGSizeEqualToSize(imageSize, viewSize) && [viewOrientation isEqualToString:dict[@"UILaunchImageOrientation"]]){
            launchImageName = dict[@"UILaunchImageName"];
            UIImage *image = [UIImage imageNamed:launchImageName];
            return image;
        }
    }
    return nil;
}

- (UIImage *)storyboardLaunchImage{
    NSString *storyboardLaunchName = [[NSBundle mainBundle].infoDictionary valueForKey:@"UILaunchStoryboardName"];
    UIViewController *launchVC = [[UIStoryboard storyboardWithName:storyboardLaunchName bundle:nil] instantiateInitialViewController];
    if(launchVC){
        UIView *view = launchVC.view;
        view.frame = [UIScreen mainScreen].bounds;
        return [self viewConvertImage:view];
    }
    return nil;
}

- (UIImage*)viewConvertImage:(UIView*)launchView{
    CGSize imageSize = launchView.bounds.size;
    UIGraphicsBeginImageContextWithOptions(imageSize, NO, [UIScreen mainScreen].scale);
    [launchView.layer renderInContext:UIGraphicsGetCurrentContext()];
    UIImage *launchImage = UIGraphicsGetImageFromCurrentImageContext();
    UIGraphicsEndImageContext();
    return launchImage;
}

- (void)dealloc {
    NSLog(@"bye----bye");
}
@end

File: ZLaunchAdSwift/ZLaunchAdSwift/ZLaunchAd/ZLaunchAdVC.swift
//
//  ZLaunchAdVC.swift
//  ZLaunchAdDemo
//
//  Created by mengqingzheng on 2017/4/5.
//  Copyright © 2017年 meng. All rights reserved.
//

import UIKit

class ZLaunchAdVC: UIViewController {
    
    fileprivate var skipBtnConfig: SkipBtnModel = SkipBtnModel()
    /// viewController display 3s without ads
    fileprivate var defaultTime = 3
    /// distance of your ad to window's bottom
    fileprivate var adViewBottomDistance: CGFloat = 100
    /// transition type when change rootViewController
    fileprivate var transitionType: TransitionType = .fade
    /// ad's show time
    fileprivate var adDuration: Int = 0
    /// original timer
    fileprivate var originalTimer: DispatchSourceTimer?
    /// ad timer
    fileprivate var dataTimer: DispatchSourceTimer?
    /// click adImg closure
    fileprivate var adImgViewClick: ZClosure?
    /// layer
    fileprivate var animationLayer: CAShapeLayer?
    fileprivate var rootViewController: UIViewController?
    /// the launch imageView
    fileprivate lazy var launchImageView: UIImageView = {
        let imgView = UIImageView(frame: UIScreen.main.bounds)
        imgView.image = self.getLaunchImage()
        return imgView
    }()
    /// your ad imgView
    fileprivate lazy var launchAdImgView: UIImageView = {
        let height = Z_SCREEN_HEIGHT - self.adViewBottomDistance
        let imgView = UIImageView(frame: CGRect(x: 0, y: 0, width: Z_SCREEN_WIDTH, height: height))
        imgView.isUserInteractionEnabled = true
        imgView.alpha = 0.2
        let tap = UITapGestureRecognizer.init(target: self, action: #selector(launchAdTapAction(sender:)))
        imgView.addGestureRecognizer(tap)
        return imgView
    }()
    /// the skip button
    fileprivate lazy var skipBtn: UIButton = {
        let button = UIButton(type: .custom)
        button.addTarget(self, action: #selector(skipBtnClick), for: .touchUpInside)
        return button
    }()
    
    /// tap your ad action
    @objc fileprivate func launchAdTapAction(sender: UITapGestureRecognizer) {
        dataTimer?.cancel()
        launchAdVCRemove {
            DispatchQueue.main.asyncAfter(deadline: DispatchTime.now() + 0.4, execute: {
                if self.adImgViewClick != nil {
                    self.adImgViewClick!()
                }
            })
        }
    }
    /// skipButton's action
    @objc fileprivate func skipBtnClick() {
        dataTimer?.cancel()
        launchAdVCRemove(completion: nil)
    }
    override func viewDidLoad() {
        super.viewDidLoad()
        view.addSubview(launchImageView)
        startTimer()
    }
    deinit {
        print("byebye")
    }
}

// MARK: - API
extension ZLaunchAdVC {
    ///
    @discardableResult
    public func defaultDuration(_ defaultDuration: Int) -> Self {
        if defaultDuration >= 1 {
            self.defaultTime = defaultDuration
        }
        return self
    }
    /// set skip button's params
    @discardableResult
    public func configSkipBtn(_ config: (inout SkipBtnModel) -> Void) -> ZLaunchAdVC {
        config(&skipBtnConfig)
        return self
    }
    /// distance of adImg to launchImage's bottom
    @discardableResult
    public func adBottom(_ adViewBottom: CGFloat) -> Self {
        adViewBottomDistance = adViewBottom
        return self
    }
    /// set transition animation type
    @discardableResult
    public func transition(_ transitionType: TransitionType) -> Self {
        self.transitionType = transitionType
        return self
    }
    /// set app's rootViewController
    @discardableResult
    public func rootVC(_ rootViewController: UIViewController) -> Self {
        self.rootViewController = rootViewController
        return self
    }
    /// set img parametes
    ///
    /// - Parameters:
    ///   - url: url
    ///   - duration: display seconds
    ///   - adImgViewClick: click ad & do something
    public func configNetImage(url: String, duration: Int, adImgViewClick: ZClosure?) {
        if url == "" { return }
        adDuration = duration < 1 ? 1 : duration
        view.addSubview(launchAdImgView)
        launchAdImgView.setImage(with: url, completion: {
            self.configSkipBtn()
            if self.originalTimer?.isCancelled == true { return }
            self.adStartTimer()
            UIView.animate(withDuration: 0.8, animations: {
                self.launchAdImgView.alpha = 1
            })
        })
        self.adImgViewClick = adImgViewClick
    }
    /// set local image
    ///
    /// - Parameters:
    ///   - image: pic name
    ///   - duration: display seconds
    ///   - adImgViewClick: do something click ad
    public func configLocalImage(image: UIImage?, duration: Int, adImgViewClick: ZClosure?) {
        if let image = image {
            adDuration = duration < 1 ? 1 : duration
            view.addSubview(launchAdImgView)
            launchAdImgView.image = image
            self.configSkipBtn()
            if self.originalTimer?.isCancelled == true { return }
            self.adStartTimer()
            UIView.animate(withDuration: 0.8, animations: {
                self.launchAdImgView.alpha = 1
            })
        }
        self.adImgViewClick = adImgViewClick
    }
    
    /// set local GIF
    ///
    /// - Parameters:
    ///   - name: pic's name
    ///   - duration: display seconds
    ///   - adImgViewClick: do something when click ad
    public func configLocalGif(name: String, duration: Int, adImgViewClick: ZClosure?) {
        adDuration = duration < 1 ? 1 : duration
        view.addSubview(launchAdImgView)
        launchAdImgView.gifImage(named: name) { 
            self.configSkipBtn()
            if self.originalTimer?.isCancelled == true { return }
            self.adStartTimer()
            UIView.animate(withDuration: 0.8, animations: {
                self.launchAdImgView.alpha = 1
            })
        }
        self.adImgViewClick = adImgViewClick
    }
}

// MARK: - configure skip button
extension ZLaunchAdVC {
    /// setup skip button
    fileprivate func configSkipBtn() -> Void {
        skipBtn.removeFromSuperview()
        if animationLayer != nil {
            animationLayer?.removeFromSuperlayer()
            animationLayer = nil
        }
        
        if skipBtnConfig.skipBtnType != .none {
            skipBtn.backgroundColor = skipBtnConfig.backgroundColor
            skipBtn.setTitleColor(skipBtnConfig.titleColor, for: .normal)
            skipBtn.titleLabel?.font = skipBtnConfig.titleFont
            
            if skipBtnConfig.skipBtnType == .circle {
                skipBtn.frame = CGRect(x: 0, y: 0, width: skipBtnConfig.height, height: skipBtnConfig.height)
                skipBtn.layer.cornerRadius = skipBtnConfig.height*0.5
                circleBtnAddLayer(strokeColor: skipBtnConfig.strokeColor, lineWidth: skipBtnConfig.lineWidth)
            } else {
                skipBtn.frame = CGRect(x: 0, y: 0, width: skipBtnConfig.width, height: skipBtnConfig.height)
                skipBtn.layer.cornerRadius = skipBtnConfig.cornerRadius
                skipBtn.layer.borderColor = skipBtnConfig.borderColor.cgColor
                skipBtn.layer.borderWidth = skipBtnConfig.borderWidth
            }
            skipBtn.center = CGPoint(x: skipBtnConfig.centerX, y: skipBtnConfig.centerY)
            skipBtn.setTitle(skipBtnConfig.skipBtnType == .timer ? "\(adDuration) 跳过" : "跳过", for: .normal)
            view.addSubview(skipBtn)
        }
    }
    
    /// circle button add animation
    fileprivate func circleBtnAddLayer(strokeColor: UIColor, lineWidth: CGFloat) {
        let bezierPath = UIBezierPath(ovalIn: skipBtn.bounds)
        animationLayer = CAShapeLayer()
        animationLayer?.path = bezierPath.cgPath
        animationLayer?.lineWidth = lineWidth
        animationLayer?.strokeColor = strokeColor.cgColor
        animationLayer?.fillColor = UIColor.clear.cgColor
        let animation = CABasicAnimation(keyPath: "strokeStart")
        animation.duration = Double(adDuration)
        animation.fromValue = 0
        animation.toValue = 1
        animationLayer?.add(animation, forKey: nil)
        skipBtn.layer.addSublayer(animationLayer!)
    }
}
// MARK: - ZLaunchAdVC remove
extension ZLaunchAdVC {
    /// remove the launch viewController, change rootViewController or do something
    fileprivate func launchAdVCRemove(completion: (()->())?) {
        if self.originalTimer?.isCancelled == false {
            self.originalTimer?.cancel()
        }
        if self.dataTimer?.isCancelled == false {
            self.dataTimer?.cancel()
        }
        guard self.rootViewController != nil else { return }
        DispatchQueue.main.asyncAfter(deadline: DispatchTime.now()+0.1, execute: {
            self.transitionAnimation()
            UIApplication.shared.keyWindow?.rootViewController = self.rootViewController
            guard completion != nil else { return }
            completion!()
        })
    }
    
    /// add tansition animation to window when viewController will dealloc
    private func transitionAnimation() {
        let trans = CATransition()
        trans.duration = 0.5
        switch transitionType {
        case .rippleEffect:
            trans.type = "rippleEffect"
        case .filpFromLeft:
            trans.type = "oglFlip"
            trans.subtype = kCATransitionFromLeft
        case .filpFromRight:
            trans.type = "oglFlip"
            trans.subtype = kCATransitionFromRight
        case .flipFromTop:
            trans.type = "oglFlip"
            trans.subtype = kCATransitionFromTop
        case .filpFromBottom:
            trans.type = "oglFlip"
            trans.subtype = kCATransitionFromBottom
        default:
            trans.type = "fade"
        }
        UIApplication.shared.keyWindow?.layer.add(trans, forKey: nil)
    }
}

//MARK: - GCD timer
/// APP启动后开始默认定时器，默认3s
/// 3s内若网络图片加载完成，默认定时器关闭，开启图片倒计时
/// 3s内若图片加载未完成，执行completion闭包
extension ZLaunchAdVC {
    /// start original timer
    fileprivate func startTimer() {
        originalTimer = DispatchSource.makeTimerSource(flags: [], queue:DispatchQueue.global())
        originalTimer?.scheduleRepeating(deadline: DispatchTime.now(), interval: DispatchTimeInterval.seconds(1), leeway: DispatchTimeInterval.milliseconds(defaultTime))
        originalTimer?.setEventHandler(handler: {
            printLog("original timer:" + "\(self.defaultTime)")
            if self.defaultTime == 0 {
                DispatchQueue.main.async {
                    self.launchAdVCRemove(completion: nil)
                }
            }
            self.defaultTime -= 1
        })
        originalTimer?.resume()
    }
    /// start your ad's timer
    fileprivate func adStartTimer() {
        if self.originalTimer?.isCancelled == false {
            self.originalTimer?.cancel()
        }
        dataTimer = DispatchSource.makeTimerSource(flags: [], queue:DispatchQueue.global())
        dataTimer?.scheduleRepeating(deadline: DispatchTime.now(), interval: DispatchTimeInterval.seconds(1), leeway: DispatchTimeInterval.milliseconds(adDuration))
        dataTimer?.setEventHandler(handler: {
            printLog("ad timer:" + "\(self.adDuration)")
            DispatchQueue.main.async {
                self.skipBtn.setTitle(self.skipBtnConfig.skipBtnType == .timer ? "\(self.adDuration) 跳过" : "跳过", for: .normal)
                if self.adDuration == 0 {
                    self.launchAdVCRemove(completion: nil)
                }
                self.adDuration -= 1
            }
        })
        dataTimer?.resume()
    }
}

// MARK: - StatusBar's Color
/// You can set statusBar by General -> Deployment Info
extension ZLaunchAdVC {
    override var prefersStatusBarHidden: Bool {
        return Bundle.main.infoDictionary?["UIStatusBarHidden"] as! Bool
    }
    override var preferredStatusBarStyle: UIStatusBarStyle {
        let str = Bundle.main.infoDictionary?["UIStatusBarStyle"] as! String
        return str.contains("Default") ? .default : .lightContent
    }
}

// MARK: - Get LaunchImage
extension ZLaunchAdVC {
    fileprivate func getLaunchImage() -> UIImage {
        if (assetsLaunchImage() != nil) || (storyboardLaunchImage() != nil) {
            return assetsLaunchImage() == nil ? storyboardLaunchImage()! : assetsLaunchImage()!
        }
        return UIImage()
    }
    
    /// From Assets
    private func assetsLaunchImage() -> UIImage? {
        let size = UIScreen.main.bounds.size
        let orientation = "Portrait" // "Landscape"
        guard let launchImages = Bundle.main.infoDictionary?["UILaunchImages"] as? [[String: Any]] else { return nil }
        for dict in launchImages {
            let imageSize = CGSizeFromString(dict["UILaunchImageSize"] as! String)
            if __CGSizeEqualToSize(imageSize, size) && orientation == (dict["UILaunchImageOrientation"] as! String) {
                let launchImageName = dict["UILaunchImageName"] as! String
                let image = UIImage.init(named: launchImageName)
                return image
            }
        }
        return nil
    }
    
    /// Form LaunchScreen.Storyboard
    private func storyboardLaunchImage() -> UIImage? {
        guard let storyboardLaunchName = Bundle.main.infoDictionary?["UILaunchStoryboardName"] as? String,
            let launchVC = UIStoryboard.init(name: storyboardLaunchName, bundle: nil).instantiateInitialViewController() else {
                return nil
        }
        let view = launchVC.view
        view?.frame = UIScreen.main.bounds
        let image = viewConvertImage(view: view!)
        return image
    }
    
    /// view convert image
    private func viewConvertImage(view: UIView) -> UIImage? {
        let size = view.bounds.size
        UIGraphicsBeginImageContextWithOptions(size, false, UIScreen.main.scale)
        view.layer.render(in: UIGraphicsGetCurrentContext()!)
        let image = UIGraphicsGetImageFromCurrentImageContext()
        UIGraphicsEndImageContext()
        return image
    }
}


